editor/create_dialog.cpp

```diff
@@ -95,364 +71,297 @@ void CreateDialog::popup_create(bool p_dont_clear, bool p_replace_mode, const St
-				if (search_box->get_text().is_subsequence_ofi(type2)) {
-					found = true;
-					break;
-				}
+float CreateDialog::_score_type(const String &p_type, const String &p_search) const {
+	// Strip off the file name from class_name scripts.
+	const String t = p_type.get_slicec(' ', 0);
 
-				type2 = cpp_type2 ? ClassDB::get_parent_class(type2) : ed.script_class_get_base(type2);
-				cpp_type2 = cpp_type2 || ClassDB::class_exists(type2); // Built-in class can't inherit from custom type, so we can skip the check if it's already true.
+	float score = _is_type_preferred(t) ? 1.f : 0.7f;
 
-				if (!cpp_type2 && !search_loaded_scripts.has(type2)) {
-					search_loaded_scripts[type2] = ed.script_class_load_script(type2);
-				}
-			}
+	// Add score for being a substring close to the start of the string.
+	float inverse_length = 1.f / float(t.length());
+	int pos = t.findn(p_search);
+	score *= (pos > -1) ? 1.f - (pos * inverse_length) : (p_search.length() * inverse_length);
 
-			if (found) {
-				add_type(type, search_options_types, root, &to_select);
-			}
-		}
+	// Favor shorter items: they resemble the search term more.
+	score *= 0.8f + 0.2f * (p_search.length() * inverse_length);
 
-		if (EditorNode::get_editor_data().get_custom_types().has(type) && ClassDB::is_parent_class(type, base_type)) {
-			//there are custom types based on this... cool.
+	// Add score for being a favorite type.
+	score *= (favorite_list.find(t) > -1) ? 1.0f : 0.7f;
 
-			const Vector<EditorData::CustomType> &ct = EditorNode::get_editor_data().get_custom_types()[type];
-			for (int i = 0; i < ct.size(); i++) {
-				bool show = search_box->get_text().is_subsequence_ofi(ct[i].name);
-
-				if (!show) {
-					continue;
-				}
-
-				if (!search_options_types.has(type)) {
-					add_type(type, search_options_types, root, &to_select);
-				}
-
-				TreeItem *ti;
-				if (search_options_types.has(type)) {
-					ti = search_options_types[type];
-				} else {
-					ti = search_options->get_root();
-				}
-
-				TreeItem *item = search_options->create_item(ti);
-				item->set_metadata(0, type);
-				item->set_text(0, ct[i].name);
-				item->set_icon(0, ct[i].icon.is_valid() ? ct[i].icon : search_options->get_theme_icon(base_icon, "EditorIcons"));
-
-				if (!to_select || ct[i].name == search_box->get_text()) {
-					to_select = item;
-				}
-			}
+	// Look through at most 5 recent items
+	bool found_recent = false;
+	for (int i = 0; i < MIN(5, recent->get_item_count()); i++) {
+		if (recent->get_item_text(i) == t) {
+			found_recent = true;
+			break;
 		}
 	}
+	score *= found_recent ? 1.0f : 0.8f;
 
-	if (search_box->get_text() == "") {
-		to_select = root;
-	}
-
-	if (to_select) {
-		to_select->select(0);
-		search_options->scroll_to_item(to_select);
-		favorite->set_disabled(false);
-		favorite->set_pressed(favorite_list.find(to_select->get_text(0)) != -1);
-	}
+	return score;
+}
 
-	get_ok()->set_disabled(root->get_children() == nullptr);
+void CreateDialog::_clean_up() {
```

Comment:
```suggestion
void CreateDialog::_cleanup() {
```
`cleanup` is quite common in the Godot codebase.